File: src/main/java/com/mxmariner/andxtidelib/remote/RemoteStationData.java
package com.mxmariner.andxtidelib.remote;

import android.os.Parcel;
import android.os.Parcelable;


public class RemoteStationData implements Parcelable {
    
    public static final int REQUEST_OPTION_PLAIN_DATA = 2;
    public static final int REQUEST_OPTION_RAW_DATA = 4;
    public static final int REQUEST_OPTION_PREDICTION = 8;
    public static final int REQUEST_OPTION_ABOUT = 16;
    public static final int REQUEST_OPTION_GRAPH_SVG = 32;
    public static final int REQUEST_OPTION_CLOCK_SVG = 64;

    protected long id;
    protected String name;
    protected String dataTimeStamp;
    protected double latitude;
    protected double longitude;

    /* optional data */
    protected String[] plainData;
    protected String[] rawData;
    protected String prediction;
    protected String about;
    protected String graphSvg;
    protected String clockSvg;

    public RemoteStationData() {
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public String getDataTimeStamp() {
        return dataTimeStamp;
    }

    /* optional data */

    public String[] getOptionalPlainData() {
        return plainData;
    }

    public String[] getOptionalRawData() {
        return rawData;
    }

    public String getOptionalPrediction() {
        return prediction;
    }

    public String getOptionalAbout() {
        return about;
    }

    public String getOptionalGraphSvg() {
        return graphSvg;
    }

    public String getOptionalClockSvg() {
        return clockSvg;
    }

    protected RemoteStationData(Parcel in) {
        id = in.readLong();
        name = in.readString();
        dataTimeStamp = in.readString();
        latitude = in.readDouble();
        longitude = in.readDouble();
        plainData = (String[]) in.readArray(String.class.getClassLoader());
        rawData = (String[]) in.readArray(String.class.getClassLoader());
        prediction = in.readString();
        about = in.readString();
        graphSvg = in.readString();
        clockSvg = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeString(name);
        dest.writeString(dataTimeStamp);
        dest.writeDouble(latitude);
        dest.writeDouble(longitude);
        dest.writeArray(plainData);
        dest.writeArray(rawData);
        dest.writeString(prediction);
        dest.writeString(about);
        dest.writeString(graphSvg);
        dest.writeString(clockSvg);
    }

    @SuppressWarnings("unused")
    public static final Parcelable.Creator<RemoteStationData> CREATOR = new Parcelable.Creator<RemoteStationData>() {
        @Override
        public RemoteStationData createFromParcel(Parcel in) {
            return new RemoteStationData(in);
        }

        @Override
        public RemoteStationData[] newArray(int size) {
            return new RemoteStationData[size];
        }
    };
}

File: mxtide/src/main/java/com/mxmariner/andxtidelib/remote/RemoteStationData.kt
package com.mxmariner.andxtidelib.remote

import android.os.Parcel
import android.os.Parcelable

const val REQUEST_OPTION_PLAIN_DATA = 2
const val REQUEST_OPTION_RAW_DATA = 4
const val REQUEST_OPTION_PREDICTION = 8
const val REQUEST_OPTION_ABOUT = 16
const val REQUEST_OPTION_GRAPH_SVG = 32
const val REQUEST_OPTION_CLOCK_SVG = 64

class RemoteStationData(val id: Long,
                        val name: String,
                        val dataTimeStamp: String,
                        val latitude: Double,
                        val longitude: Double,
                        val stationType: StationType,
                        var optionalPlainData: Array<String>? = null,
                        var optionalRawData: Array<String>? = null,
                        var optionalPrediction: String? = null,
                        var optionalAbout: String? = null,
                        var optionalGraphSvg: String? = null,
                        var optionalClockSvg: String? = null) : Parcelable {

    constructor(parcel: Parcel) : this(
            parcel.readLong(),
            parcel.readString(),
            parcel.readString(),
            parcel.readDouble(),
            parcel.readDouble(),
            parcel.readParcelable(StationType::class.java.classLoader),
            parcel.createStringArray(),
            parcel.createStringArray(),
            parcel.readString(),
            parcel.readString(),
            parcel.readString(),
            parcel.readString())

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeLong(id)
        parcel.writeString(name)
        parcel.writeString(dataTimeStamp)
        parcel.writeDouble(latitude)
        parcel.writeDouble(longitude)
        parcel.writeParcelable(stationType, flags)
        parcel.writeStringArray(optionalPlainData)
        parcel.writeStringArray(optionalRawData)
        parcel.writeString(optionalPrediction)
        parcel.writeString(optionalAbout)
        parcel.writeString(optionalGraphSvg)
        parcel.writeString(optionalClockSvg)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<RemoteStationData> {

        override fun createFromParcel(parcel: Parcel): RemoteStationData {
            return RemoteStationData(parcel)
        }

        override fun newArray(size: Int): Array<RemoteStationData?> {
            return arrayOfNulls(size)
        }
    }
}
